client/src/components/side-bar.tsx

```diff
@@ -0,0 +1,614 @@
+    padding: theme.spacing(0, 1),
+    // necessary for content to be below app bar
+    ...theme.mixins.toolbar,
+  }));
+
+  const [drawerOpen, setDrawerOpen] = React.useState(true);
+
+  const handleDrawerChange = () => {
+    setDrawerOpen(!drawerOpen);
+    console.log("button clicked");
+  };
+
+  const Drawer = styled(MuiDrawer, {
+    shouldForwardProp: (prop) => prop !== "open",
+  })(({ theme, open }) => ({
+    width: drawerWidth,
+    flexShrink: 0,
+    whiteSpace: "nowrap",
+    boxSizing: "border-box",
+    ...(open && {
+      ...openedMixin(theme),
+      "& .MuiDrawer-paper": openedMixin(theme),
+    }),
+    ...(!open && {
+      ...closedMixin(theme),
+      "& .MuiDrawer-paper": closedMixin(theme),
+    }),
+  }));
+
+  const [shareOpen, setShareOpen] = React.useState(false);
+  const handleClickOpenShare = () => {
+    setShareOpen(true);
+  };
+
+  const handleCloseShare = () => {
+    setShareOpen(false);
+  };
+
+  function isExpValid(exp: LessonExpectation): boolean {
+    if (!exp.features) {
+      return true;
+    }
+    return validateExpectationFeatures(exp.features);
+  }
+
+  function handleLaunch() {
+    saveChanges();
+    const host = process.env.TUTOR_ENDPOINT || location.origin;
+    const guest = `&guest=${context.user?.name}`;
+    const path = `${host}/tutor?lesson=${lessonId}&admin=true${guest}`;
+    window.location.href = path;
+  }
+  function isLessonValid(): boolean {
+    if (!lessonUnderEdit.lesson) {
+      return false;
+    }
+    return (
+      !error &&
+      lessonUnderEdit.lesson.arch != OPENAI_CLASSIFIER_ARCHITECTURE &&
+      lessonUnderEdit.lesson?.expectations.every((exp: LessonExpectation) =>
+        isExpValid(exp)
+      )
+    );
+  }
+
+  function handleDiscard() {
+    navigate(`/lessons`);
+  }
+
+  function handleSavePopUp(open: boolean): void {
+    setSavePopUp(open);
+  }
+
+  function handleSaveExit(): void {
+    saveChanges();
+    navigate(`/lessons`);
+  }
+
+  function handleSaveContinue(): void {
+    saveChanges();
+    handleSavePopUp(false);
+  }
+
+  function saveChanges(): void {
+    if (!lessonUnderEdit.lesson) {
+      return;
+    }
+    toast("Saving...");
+    const convertedLesson: Lesson = { ...lessonUnderEdit.lesson };
+    if (!lessonId) {
+      convertedLesson.createdBy = context.user?.id || "";
+    }
+    const origId = lessonId || lessonUnderEdit.lesson?.lessonId;
+    updateLesson(origId, convertedLesson, cookies.accessToken)
+      .then((lesson) => {
+        if (lesson) {
+          setLesson(lesson);
+        }
+
+        if (lessonId !== lesson?.lessonId) {
+          // window.location.href = `/lessons/edit?lessonId=${lesson.lessonId}`;
+          setLessonId(lesson.lessonId);
+          // window.location.reload();
+          navigate("/lessons");
+        }
+        toast("Success!");
+      })
+      .catch((err) => {
+        toast("Failed to save lesson.");
+        console.error(err);
+      });
+  }
+
+  const trainAIButtonColor =
```

Comment:
move this into a helper function in a separate helper file, this code should reduce down to something like:

`const trainAiButtonColor = getTrainButtonColor(trainStatus)`